File: luaSource/unit_test/test_call_selector_argument.lua
-- 测试是否传值给oc，判断都在oc中进行

_G.TestCallSelectorArgument_block1 = function (instance,i)
	-- body
	print("TestCallSelectorArgument_block1 i")
	print(i)
	return i + 1
end

_G.TestCallSelectorArgument_block2 = function (instance,str)
	-- body
	print("TestCallSelectorArgument_block2 str")
	print(str)
	-- return i + 1
end


describe("test call selector argument",function ( ... )
	-- body
	local luapatch = require("luapatch")
	local packClass = luapatch.packClass
	local TestCallSelectorArgument = packClass("TestCallSelectorArgument")
	local NSDictionary = packClass("NSDictionary")
	it("pass c value to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkInt(1)
		TestCallSelectorArgument.checkDouble(1.1)
	end)

	it("pass string and num to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkNSString("hello world")
		TestCallSelectorArgument.checkNSNumber(1)
	end)

	it("pass struct to oc",function ( ... )
		-- body
		-- 在测试结构体的那里已经测试了
	end)

	it("pass id to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkIsDictionary(NSDictionary.new())
	end)

	it("pass class obj to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkIsDictionaryClass(NSDictionary.class())
	end)

	it("pass char * to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkCStr("hello world")
	end)

	it("pass sel to oc",function ( ... )
		-- body
		TestCallSelectorArgument.checkSEL("checkSEL")
	end)

	it("pass block to oc",function ( ... )
		-- body
        local block = luapatch.convertLuaBlockToObjcBlock("TestCallSelectorArgument_block1","int_int")
        TestCallSelectorArgument.checkBlock(block)
        local block2 = luapatch.convertLuaBlockToObjcBlock("TestCallSelectorArgument_block2","void_id")
		TestCallSelectorArgument.checkBlock2(block2)
	end)
end)

File: luaPatchDemo/TestCallSelectorArgument.m
//
//  TestLuaToOCValue.m
//  luaPatchDemo
//
//  Created by 黄钊 on 2019/9/7.
//  Copyright © 2019 hz. All rights reserved.
//

#import "TestCallSelectorArgument.h"
#import <UIKit/UIKit.h>
#include "string.h"
//测试lua传递参数给oc

static BOOL doubleEqual(double a,double b){
    double x = a - b;
    const double EPSINON = 0.000001;
    return (x >= -EPSINON)&&(x <= EPSINON);
}

@implementation TestCallSelectorArgument

+(void)checkInt:(int)i{
    NSAssert(i == 1,@"TestLuaToOCValue checkInt value error");
}

+(void)checkDouble:(double)d{
    NSAssert(doubleEqual(d,1.1),@"TestLuaToOCValue checkDouble value error");
}

+(void)checkNSString:(NSString *)str{
    NSAssert([str isEqualToString:@"hello world"],@"TestLuaToOCValue checkNSString value error");
}

+(void)checkNSNumber:(NSNumber *)num{
    NSAssert(num.intValue == 1,@"TestLuaToOCValue checkNSNumber value error");
}

+(void)checkOCStructWithRect:(CGRect)rect Point:(CGPoint)point Size:(CGSize)size Range:(NSRange)range{
    NSAssert(rect.origin.x == 1 && rect.origin.y == 2 && rect.size.width == 3 && rect.size.height == 4, @"TestLuaToOCValue checkOCStruct rect error");
    NSAssert(point.x == 1 && point.y == 2,@"TestLuaToOCValue checkOCStruct point error");
    NSAssert(size.width == 1 && size.height == 2,@"TestLuaToOCValue checkOCStruct size error");
    NSAssert(range.location == 1 && range.length == 2,@"TestLuaToOCValue checkOCStruct range error");
}

+(void)checkIsDictionary:(id)i{
    NSAssert([i isKindOfClass:NSDictionary.class],@"TestLuaToOCValue checkIsDictionary error");
}

+(void)checkIsDictionaryClass:(id)i{
    NSAssert(i == NSDictionary.class,@"TestLuaToOCValue checkIsDictionaryClass error");
}

+(void)checkCStr:(char *)str{
    NSAssert(strcmp(str, "hello world") == 0,@"TestLuaToOCValue checkCStr error");
}

+(void)checkSEL:(SEL)sel{
    NSAssert([NSStringFromSelector(sel) isEqualToString:@"checkSEL"],@"TestLuaToOCValue checkSEL error");
}

+(void)checkBlock:(int (^)(int))block{
    int ret = block(1);
    NSAssert(ret == 2, @"TestLuaToOCValue checkBlock error");
}

+(void)checkBlock2:(void (^)(NSString*))block{
    block(@"hello");
}

@end
